python/cudf/cudf/core/reshape.py

```diff
@@ -248,10 +264,27 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
             f'`axis` must be 0 / "index" or 1 / "columns", got: {axis}'
         )
 
+    # Retrieve the base types of `objs`. In order to support sub-types
+    # and object wrappers, we use `isinstance()` instead of comparing
+    # types directly
+    typs = set()
+    for o in objs:
+        if isinstance(o, cudf.MultiIndex):
+            typs.add(cudf.MultiIndex)
+        elif isinstance(o, cudf.BaseIndex):
+            typs.add(type(o))
+        elif isinstance(o, cudf.DataFrame):
+            typs.add(cudf.DataFrame)
+        elif isinstance(o, cudf.Series):
+            typs.add(cudf.Series)
+        else:
+            raise TypeError(f"cannot concatenate object of type {type(o)}")
+
+    allowed_typs = {cudf.Series, cudf.DataFrame}
```

Comment:
*nit*: This is a slight misnomer because we do allow indices as input as well if and only if we provide only indices. My preferred solution would be to disallow concatenating indices (as pandas does), but perhaps that breaks too many other things.

Reply:
simplified this portion of the code and included an `assert` for index-exclusive concatenating